src/program/config/xpath_printer.lua

```diff
@@ -0,0 +1,97 @@
+module(..., package.seeall)
+
+Parser = {}
+
+function Parser.new (text)
+   local o = {
+      text = assert(text),
+      pos = 1,
+      length = #text,
+   }
+   return setmetatable(o, {__index = Parser})
+end
+
+function Parser:skip_whitespace ()
+   local pos = self.text:match("^[ \t]+()", self.pos)
+   if pos then self.pos = pos end
+end
+
+function Parser:token ()
+   self:skip_whitespace()
+   local token, pos = self.text:match("([^ \n\t;]+)()", self.pos)
+   self.pos = pos
+   return token
+end
+
+function Parser:peep ()
+   self:skip_whitespace()
+   return self.text:match("([^ \n\t;]+)", self.pos)
+end
+
+function Parser:consume (str)
+   local tok, pos = self.text:match("([^ \n\t;]+)()", self.pos)
+   assert(tok == str)
+   self.pos = pos
+end
+
+function Parser:eof ()
+   return not self:peep()
+end
+
+XPathPrinter = {}
+
+function XPathPrinter.new (text)
+   local o = {
+      p = Parser.new(text),
+   }
+   return setmetatable(o, {__index = XPathPrinter})
+end
+
+function XPathPrinter:parse ()
+   local p = self.p
+   local path, output = {''}, {}
+   local function parse (path, output)
+      if p:eof() then return end
+      local tok = p:token()
+      if tok == '}' then
+         table.remove(path)
+      elseif p:peep() == '{' then
```

Comment:
This isn't going to work in general.  The format for the xpath has one component per tree, yes, but for lists for example you are going to have e.g. `/foo/bar[a=A][b=B]/qux'.  I think doing this transformation on the client side is the wrong thing; we need to tell the leader that we want output in xpath format and have it generate it, using a printer derived from the grammar.  That way we also avoid paging the whole thing into memory on the client side.